Docs: new tutorial on converting a frontend data source (#915)
grafana/plugin-tools: docusaurus/docs/tutorials/convert-a-frontend-datasource-to-backend.md

## Key concepts

Before going into specific conversion advice, we will discuss the main components of a data source and how these differ between frontend and backend plugins.

### Frontend `DataSource` class

Data source plugins implement a new `DataSourcePlugin`. This class takes as a parameter a `DataSource` class, which for frontend data sources extends `DataSourceApi`, and for backend data sources extends `DataSourceWithBackend`. Because the `DatasourceWithBackend` class already implements most of the required methods, you can migrate to it to significantly simplify your code.

Data source plugins require two components: a query editor and a config editor.

**Examples:**

- [Frontend data source](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/datasource-http/src/DataSource.ts#L14).
- [Backend data source](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/datasource-http-backend/src/datasource.ts#L6).

> ### Query and config editor

Author: FTR, I have merged the query and config editor sections into one since it was quite similar.

keywords:
  - grafana
  - plugins
  - plugin
  - backend
  - frontend
  - datasource
---

# How to convert a frontend data source plugin into a backend plugin

This guide shows you how to convert an existing frontend-only data source plugin into a [backend plugin](../introduction/backend.md).

To convert the frontend data source, we recommend scaffolding a new backend data source plugin using [`npx @grafana/create-plugin@latest`](https://grafana.com/developers/plugin-tools/). Use the following instructions to extend this foundation to copy functionality from your original plugin.

## Why

There are multiple features available only in backend plugins, such as Grafana Alerting or Recorded queries. See the use cases for implementing a backend plugin in the [backend plugins introduction](../introduction/backend.md#use-cases-for-implementing-a-backend-plugin).

## Before you begin

Before you dive into the details, you should familiarize yourself with the process of creating a backend data source plugin. If you haven't done this before, you can follow our tutorial for [building a backend plugin](./build-a-data-source-backend-plugin.md).

## Key concepts

> Before going into specific conversion advice, we will discuss the main components of a data source and how these differ between frontend and backend plugins.

Reviewer: ```suggestion
Before going into specific conversion advice, it's important to understand the main components of a data source and how these differ between frontend and backend plugins.
```

### Frontend `DataSource` class

Data source plugins implement a new `DataSourcePlugin`. This class takes as a parameter a `DataSource` class, which for frontend data sources extends `DataSourceApi`, and for backend data sources extends `DataSourceWithBackend`. Because the `DatasourceWithBackend` class already implements most of the required methods, you can migrate to it to significantly simplify your code.

Data source plugins require two components: a query editor and a config editor.

**Examples:**

- [Frontend data source](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/datasource-http/src/DataSource.ts#L14).
- [Backend data source](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/datasource-http-backend/src/datasource.ts#L6).

### Query and config editor

> These two frontend components do not require to be changed when converting a frontend data source to a backend data source. However, if you add a backend component to a data source you can request `resources` from it. We will talk more about this in the [resource requests section](#other-resource-requests) but resources are additional endpoints that the plugin exposes and can be used to populate or validate the query or config editor.

Reviewer: ```suggestion
These two frontend components do not need to be changed when converting a frontend data source to a backend data source. However, if you add a backend component to a data source you can request `resources` from it.  Resources are additional endpoints that the plugin exposes and can be used to populate or validate the query or config editor. Learn about this in the [resource requests section](#other-resource-requests).
```

In a frontend-only data source, any request that requires authentication needs to go through the plugin proxy. You need to define a `routes` object within the `plugin.json` file and specify there the URL and credentials to use for each request. For example, you can authenticate a request to a given URL by setting an `Authorization` header with the `SecureJsonData` credentials:

```json title="src/plugin.json"
"routes": [
  {
    "path": "example",
    "url": "https://api.example.com",
    "headers": [
      {
        "name": "Authorization",
        "content": "Bearer {{ .SecureJsonData.apiToken }}"
      }
    ]
  }
]
```

To use this route, the frontend data source should call the `fetch` method from the `DataSourceApi` class. This method proxies the request and adds the `Authorization` header:

```typescript title="src/DataSource.ts"
import { getBackendSrv } from '@grafana/runtime';

const routePath = '/example';

const res = getBackendSrv().datasourceRequest({
  url: this.url + routePath + '/v1/users',
  method: 'GET',
});
// Handle response
```

> In a backend data source, the authentication logic should be moved to the `Datasource` constructor. This method is called when the data source is created and should be used to create the authenticated client. Store this client in the `Datasource` instance and use it for every request. For example:

Reviewer: ```suggestion
In a backend data source, you should move the authentication logic to the `Datasource` constructor. This method is called when the data source is created and should be used to create the authenticated client. Store this client in the `Datasource` instance and use it for every request. For example:
```

```go title="pkg/plugin/datasource.go"
func NewDatasource(ctx context.Context, settings backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
	opts, err := settings.HTTPClientOptions(ctx)
	if err != nil {
		return nil, fmt.Errorf("http client options: %w", err)
	}

	cl, err := httpclient.New(opts)
	if err != nil {
		return nil, fmt.Errorf("httpclient new: %w", err)
	}

return &Datasource{
		settings:   settings,
		httpClient: cl,
	}, nil
}

func (d *Datasource) CheckHealth(ctx context.Context, _ *backend.CheckHealthRequest) (*backend.CheckHealthResult, error) {
	resp, err := d.httpClient.Get(d.settings.URL + "/v1/users")
	if err != nil {
    // Log the error here
  	return &backend.CheckHealthResult{
	  	Status: backend.HealthStatusError,
		  Message: "request error",
	  }, nil
	}
	if resp.StatusCode != http.StatusOK {
  	return &backend.CheckHealthResult{
	  	Status: backend.HealthStatusError,
		  Message: fmt.Sprintf("got response code %d", resp.StatusCode),
	  }, nil
	}
	return &backend.CheckHealthResult{
		Status:  backend.HealthStatusOk,
		Message: "Data source is working",
	}, nil
}
```

:::note

> This example covers an HTTP-only data source. So, if your data source requires a database connection, you can use the Golang client for the database and execute a simple query like `SELECT 1` or a `ping` function.

Reviewer: ```suggestion
This example covers an HTTP-only data source. So, if your data source requires a database connection, you can use the Go client for the database and execute a simple query like `SELECT 1` or a `ping` function.
```

:::

### Query

> The next step is to move the query logic. This will highly vary depending on how the plugin queries the data source and transforms the response into [frames](../introduction/data-frames.md). In this tutorial, let's see how we can migrate a simple example.

Reviewer: ```suggestion
The next step is to move the query logic. This will significantly vary depending on how the plugin queries the data source and transforms the response into [frames](../introduction/data-frames.md). In this guide, you'll see how to migrate a simple example.
```

```go title="pkg/plugin/datasource.go"
func NewDatasource(_ context.Context, _ backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
	return &Datasource{
		CallResourceHandler: newResourceHandler(),
	}, nil
}

func newResourceHandler() backend.CallResourceHandler {
	mux := http.NewServeMux()
	mux.HandleFunc("/tables", handleTables)

	return httpadapter.New(mux)
}

func handleTables(w http.ResponseWriter, r *http.Request) {
  // Get tables
  res, err :=	http.DefaultClient.Get("https://api.example.com/api/v1/tables")
  // Handle errors (omited)
	body, err := io.ReadAll(res.Body)
  // Handle errors (omited)

	w.Write(body)
	w.WriteHeader(http.StatusOK)
}
```

To request the resource in the frontend, you can use the methods exposed in the base class `DataSourceWithBackend` (for example, `getResource` or `postResource`):

```typescript title="src/DataSource.ts"
export class DataSource extends DataSourceWithBackend<MyQuery, MyDataSourceOptions> {
  async getTables() {
    const response = await this.getResource('tables');
    return response;
  }
}
```

## Conclusion

> This tutorial covered the main steps to convert a frontend data source into a backend data source. There is a wide variety of plugins so if you have any questions or need help with a specific case, we encourage you to reach out in our [Community forum](https://community.grafana.com/c/plugin-development/30). Contributions to this guide are also welcome.

Reviewer: ```suggestion
This tutorial covered the main steps to convert a frontend data source into a backend data source. There are a wide variety of plugins so if you have any questions or need help with a specific case, we encourage you to reach out in our [Community forum](https://community.grafana.com/c/plugin-development/30). Contributions to this guide are also welcome.
```

---

> id: convert-a-frontend-datasource-to-backend

Reviewer: This is okay for now, but this will probably go in the /how-to-guides section instead of /tutorials after the reorg.

title: How to convert a frontend data source plugin into a backend plugin
sidebar_position: 10
description: Learn how to convert a frontend data source plugin into a backend plugin
keywords:
  - grafana
  - plugins
  - plugin
  - backend
  - frontend
  - datasource
---

# How to add a frontend data source to a backend component

This guide shows you how to convert an existing frontend-only data source plugin into a [backend plugin](https://grafana.com/developers/plugin-tools/introduction/backend-plugins).

To convert the frontend data source, we recommend scaffolding a new backend data source plugin using []`npx @grafana/create-plugin@latest`](https://grafana.com/developers/plugin-tools/). Use the following instructions to extend this foundation to copy functionality from your original plugin.

> ## Before you begin

Reviewer: are there any other prerequisites not mentioned?

> Before you dive into the details, you should familiarize yourself with the process of creating a backend data source plugin. If you haven't done this before, you can follow our tutorial for [building a backend plugin](https://grafana.com/developers/plugin-tools/tutorials/build-a-data-source-backend-plugin).

Reviewer: I would like to see a paragraph with a list of example use cases added near the top of this guide.

Author: sorry, bad wording from my side, I wanted to say a wide variety of plugins. I have rephrased that but also added a small section here for the use cases for a bakend plugin

```go title="pkg/plugin/datasource.go"
func NewDatasource(_ context.Context, _ backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
	return &Datasource{
		CallResourceHandler: newResourceHandler(),
	}, nil
}

func newResourceHandler() backend.CallResourceHandler {
	mux := http.NewServeMux()
	mux.HandleFunc("/tables", handleTables)

	return httpadapter.New(mux)
}

func handleTables(w http.ResponseWriter, r *http.Request) {
  // Get tables
  res, err :=	http.DefaultClient.Get("https://api.example.com/api/v1/tables")
  // Handle errors (omited)
	body, err := io.ReadAll(res.Body)
  // Handle errors (omited)

	w.Write(body)
	w.WriteHeader(http.StatusOK)
}
```

To request the resource in the frontend, you can use the methods exposed in the base class `DataSourceWithBackend` (for example, `getResource` or `postResource`):

```typescript title="src/DataSource.ts"
export class DataSource extends DataSourceWithBackend<MyQuery, MyDataSourceOptions> {
  async getTables() {
    const response = await this.getResource('tables');
    return response;
  }
}
```

> 

Reviewer: are there any trouble spots or commonly encountered problems or bugs that the plugin developer may face? if so, add a section to address them

Author: not that I know of
